web/src/js/components/eventlog.js

```diff
@@ -1,115 +1,151 @@
-        var filterFn = function (entry) {
-            return this.props.filter[entry.level];
-        };
-        var view = new StoreView(this.context.eventStore, filterFn.bind(this));
-        view.addListener("add", this.onEventLogChange);
-        view.addListener("recalculate", this.onEventLogChange);
+    componentDidMount() {
+        window.addEventListener("resize", this.onViewportUpdate);
+        this.view.addListener("add", this.onChange);
+        this.view.addListener("recalculate", this.onChange);
+        this.onViewportUpdate();
+    }
 
-        return {
-            view: view
-        };
-    },
-    componentWillUnmount: function () {
-        this.state.view.close();
-    },
-    filter: function (entry) {
-        return this.props.filter[entry.level];
-    },
-    onEventLogChange: function () {
-        this.forceUpdate();
-    },
-    componentWillReceiveProps: function (nextProps) {
+    componentWillUnmount() {
+        window.removeEventListener("resize", this.onViewportUpdate);
+        this.view.removeListener("add", this.onChange);
+        this.view.removeListener("recalculate", this.onChange);
+        this.view.close();
+    }
+
+    componentDidUpdate() {
+        this.onViewportUpdate();
+    }
+
+    componentWillReceiveProps(nextProps) {
         if (nextProps.filter !== this.props.filter) {
-            this.props.filter = nextProps.filter; // Dirty: Make sure that view filter sees the update.
-            this.state.view.recalculate();
+            this.view.recalculate(
+                entry => nextProps.filter[entry.level]
+            );
         }
-    },
-    getDefaultProps: function () {
-        return {
-            rowHeight: 45,
-            rowHeightMin: 15,
-            placeholderTagName: "div"
-        };
-    },
-    renderRow: function (elem) {
-        return <LogMessage key={elem.id} entry={elem}/>;
-    },
-    render: function () {
-        var entries = this.state.view.list;
-        var rows = this.renderRows(entries);
-
-        return <pre onScroll={this.onScroll}>
-            { this.getPlaceholderTop(entries.length) }
-            {rows}
-            { this.getPlaceholderBottom(entries.length) }
-        </pre>;
     }
-});
 
-var ToggleFilter = React.createClass({
-    toggle: function (e) {
-        e.preventDefault();
-        return this.props.toggleLevel(this.props.name);
-    },
-    render: function () {
-        var className = "label ";
-        if (this.props.active) {
-            className += "label-primary";
-        } else {
-            className += "label-default";
+    onViewportUpdate() {
+        const viewport = ReactDOM.findDOMNode(this);
+
+        const vScroll = calcVScroll({
+            itemCount: this.state.entries.length,
+            rowHeight: this.props.rowHeight,
+            viewportTop: viewport.scrollTop,
+            viewportHeight: viewport.offsetHeight,
+            itemHeights: this.state.entries.map(entry => this.heights[entry.id]),
```

Comment:
@mhils The component behaves weird if there are some large chunks in it. I think this mechanism is necessary for a better user expeirence. As for the performance issue, it works smoothly on my laptop, and it can be faster if we do some optimization on it. Anyway, I'll remove this if you still think it's unnecessary, but I think it's really great for UX.